### PR TITLE
Add HTTP API to main menu

### PR DESCRIPTION
**Pull request by rubenwardy**
Adds support for the HTTP API to the main menu Lua environment.

This does not include support for the `api.fetch` as that requires `minetest.after`

## How to test

Testing

### CODE REVIEW

File: src/script/scripting_mainmenu.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
 
+#include <script/lua_api/l_http.h>
```
Comment (sfan5) [inline src/script/scripting_mainmenu.cpp:20]:
```suggestion
#include "script/lua_api/l_http.h"
```

File: doc/menu_lua_api.txt [hunk 2]
```diff
@@ -67,7 +80,90 @@ core.get_mapgen_names([include_hidden=false]) -> table of map generator algorith
...
+* core.download_file(url, target) (possible in async calls)
+    * url to download, and target to store to
+    * returns true/false
+* `minetest.request_http_api()` (possible in async calls)
+    * returns `HTTPApiTable` containing http functions if the calling mod has
+      been granted access by being listed in the `secure.http_mods` or
+      `secure.trusted_mods` setting, otherwise returns `nil`.
+    * The returned table contains the functions `fetch`, `fetch_async` and
+      `fetch_async_get` described below.
+    * Only works at init time and must be called from the mod's main scope
```
Comment (sfan5) [inline doc/menu_lua_api.txt:96]: inaccurate

File: src/script/lua_api/l_http.cpp [hunk 3]
```diff
@@ -172,11 +195,15 @@ int ModApiHttp::l_request_http_api(lua_State *L)
...
 	HTTP_API(fetch_async_get);
-
-	// Stack now looks like this:
-	// <core.http_add_fetch> <table with fetch_async, fetch_async_get>
-	// Now call core.http_add_fetch to append .fetch(request, callback) to table
-	lua_call(L, 1, 1);
+	if (isMainmenu)
+		HTTP_API(fetch_sync);
+
+	if (lua_isfunction(L, -2)) {
```
Comment (sfan5) [inline src/script/lua_api/l_http.cpp:201]: would be better to just implement `core.http_add_fetch` for the menu

File: src/script/lua_api/l_http.cpp [hunk 4]
```diff
@@ -83,6 +83,25 @@ void ModApiHttp::push_http_fetch_result(lua_State *L, HTTPFetchResult &res, bool
...
+	read_http_fetch_request(L, req);
+
+	infostream << "Mod performs HTTP request with URL " << req.url << std::endl;
+
+	HTTPFetchResult res;
+	httpfetch_sync(req, res);
+
+	push_http_fetch_result(L, res, true);
+
+	// Co
```
Comment (sfan5) [inline src/script/lua_api/l_http.cpp:101]:
```suggestion
```

File: builtin/mainmenu/dlg_contentstore.lua [hunk 5]
```diff
@@ -17,6 +17,8 @@
 
 local store = { packages = {}, packages_full = {} }
 
+local http = minetest.request_http_api()
```
Comment (sfan5) [inline builtin/mainmenu/dlg_contentstore.lua:20]: does this return nil if not compiled with curl?
if yes that needs to be handled

**Comment by rubenwardy** [comment]
image

File: src/script/lua_api/l_http.cpp [hunk 6]
```diff
@@ -180,11 +198,34 @@ int ModApiHttp::l_request_http_api(lua_State *L)
...
+}
+
 #endif
 
 void ModApiHttp::Initialize(lua_State *L, int top)
 {
 #if USE_CURL
-	API_FCT(request_http_api);
+
+#ifndef SERVER
```
Comment (sfan5) [inline src/script/lua_api/l_http.cpp:220]: readability suggestion:
```c++
    bool is_mainmenu = false;
#ifndef SERVER
    is_mainmenu = !!ModApiBase::getGuiEngine(L);
#endif 
    if (is_mainmenu)
    [...]
```

File: src/script/lua_api/l_http.cpp [hunk 7]
```diff
@@ -180,11 +198,35 @@ int ModApiHttp::l_request_http_api(lua_State *L)
...
 #endif
 
 void ModApiHttp::Initialize(lua_State *L, int top)
 {
 #if USE_CURL
-	API_FCT(request_http_api);
+
+	bool isMainmenu = false;
+#ifndef SERVER
+	isMainmenu = ModApiBase::getGuiEngine(L);
```
Comment (sfan5) [inline src/script/lua_api/l_http.cpp:222]: surprised this doesn't raise a warning, I'd prefer an explicit `!!` or ` != nullptr`.

**Pull request merged**